### PR TITLE
cgroups / lxc / docker performance monitoring

### PR DESCRIPTION
**Pull request by ktsaou**
This PR adds a new data collection module, that allows netdata to collect cgroups resource utilization:

The following are collected **for each cgroup**:

- [x] `cpuacct.stat`, that provides CPU utilization (user / system)
- [x] `cpuacct.usage_percpu`, that provides CPU utilization per core
- [x] `memory.stat`, that provides memory utilization and memory activity metrics - the code that collects hierarchical data is commented. Only the values of the current container are collected.
- [x] `blkio.io_service_bytes`, `blkio.io_serviced`, `blkio.throttle.io_service_bytes`, `blkio.throttle.io_serviced`, `blkio.io_merged`, `blkio.io_queued`. All these provide also metrics per block device, but the plugin just aggregates all the disks together (no per disk information).

---

## Key features

- By default the plugin will not collect values for the host cgroup `/`, the `/lxc` and `/docker` parent cgroups (the lxc and docker sub-containers are collected).

- The plugin will by default scan for new cgroups every 10 seconds.

- The plugin has an internal protection for cases new cgroups are constantly being created and erased:
    - The setting `enable new cgroups detected at run time = yes` can be set to `no` to prevent netdata from collecting data for containers not specifically enabled at its configuration.
    - The setting `max cgroups to allow = 50` will stop netdata from continuously adding new cgroup charts. If it reaches 50 cgroups it will stop. Of course, if you need more than 50, just increase this setting.
    - The setting `max cgroups depth to monitor = 0` can be set to a positive number to limit the depth in the cgroups hierarchy that netdata will look for new containers.

---

## Charts

The above provide the following charts **per cgroup**:

1. CPU Usage
2. CPU Usage Per Core
3. Memory Usage
4. Writeback Memory
5. Memory Activity (pages in/out in MB)
6. Memory Pagefaults (major page faults = swap activity)
7. I/O Bandwidth (all disks)
8. Serviced I/O Operations (all disks)
9. Throttle I/O Bandwidth (all disks)
10. Throttle Serviced I/O Operations (all disks)
11. Queued I/O Operations (all disks)
12. Merged I/O Operations (all disks)

All the charts that have **zero values** (except  Queued I/O Operations (all disks)), will not be rendered (and memory will not be allocated for them at the netdata server).

## Current Issues

- The plugin has hardcoded the `/sys/fs/cgroup` path. It can be changed from `netdata.conf` but it should use `/proc/self/mountinfo` for automatically detecting the proper paths.
- Docker names are collected with an external script, that just shortens the docker long ids. It supports manual configuration for setting proper docker names.

### CODE REVIEW

File: src/common.c [hunk 1]
```diff
@@ -281,8 +281,273 @@ unsigned char netdata_keys_map[256] = {
...
+		[83] = 's', // S
+		[84] = 't', // T
+		[85] = 'u', // U
+		[86] = 'v', // V
+		[87] = 'w', // W
+		[88] = 'x', // X
+		[89] = 'y', // Y
+		[90] = 'z', // Z
+		[91] = '_', // [
+		[92] = '/', // backslash
```
Comment (glensc) [inline src/common.c]: backslash can be written as '\\' in C source :)

File: src/sys_fs_cgroup.c [hunk 2]
```diff
@@ -0,0 +1,1021 @@
...
+	unsigned long long cache;
+	unsigned long long rss;
+	unsigned long long rss_huge;
+	unsigned long long mapped_file;
+	unsigned long long writeback;
+	unsigned long long pgpgin;
+	unsigned long long pgpgout;
+	unsigned long long pgfault;
+	unsigned long long pgmajfault;
+/*	unsigned long long inactive_anon;
```
Comment (glensc) [inline src/sys_fs_cgroup.c]: minor ocd: i would put the `/*` on separate line, so when enabling the line, the actual line with code is not changed on that changeset, i.e `git annotate`

File: src/common.c (hunk 1 above)
Comment (ktsaou) [inline src/common.c]: Clion, my IDE does not like it.

File: src/sys_fs_cgroup.c (hunk 2 above)
Comment (ktsaou) [inline src/sys_fs_cgroup.c]: sure you are right!

**Comment by ktsaou** [comment]
I read on various blogs that `memory.stat` provides a metric for `swap`. Mine does not have it. Do you have it? Can you post a `memory.stat` file?

**Comment by Kubuxu** [comment]
Here is mine: https://gist.github.com/Kubuxu/4bd4f38fa7303e1d42a43110eb407154

**Comment by ktsaou** [comment]
ok. added dirty and swap. I guess these are control by a kernel setting.

I read that the kernel has to be boot with `cgroup_enable=memory swapaccount=1` for these to be present: https://blog.docker.com/2013/10/gathering-lxc-docker-containers-metrics/

**Comment by ktsaou** [comment]
as discussed here: https://github.com/firehol/netdata/issues/115#issuecomment-214042069 :

1. cgroups data collection is now running on its own thread
2. cgroups names are collected via an external script `cgroup-name.sh` which uses the file `/etc/netdata/cgroups-names.conf` to rename the charts.

 The format of this file is very simple: 2 columns with space between them: the first is the id of the cgroup as found in `/sys` and the second the id/name you want it to have, something like:

```
lxc/centos7 centos7
docker/<LONGID> mydocker1
```

So, what remains is to automatically collect mount info for auto-detecting where to look for cgroup files.

This will be solved later with #295 that requires the same mount info. For the moment if your system mounts cgroups anywhere else except `/sys/fs/cgroup`, just edit `netdata.conf` and set the proper cgroup paths.

So, I believe this is ready to be merged.

**Comment by ktsaou** [comment]
all current issues are resolved. updated description.

**Pull request merged**